### PR TITLE
Enable/disable a specific rule

### LINKED ISSUE #1084: Add delegate to SonarAnalysisContext to switch individual rules on/off
See [MMF-1048](https://jira.sonarsource.com/browse/MMF-1048).
In Connected Mode, SonarLint needs to be able to programmatically enable and disable rules to match the current Quality Profile.

The current _ShouldAnalysisBeDisabled_ rule is not sufficiently fine-grained as it does not allow individual rules to be turned on or off.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #1084

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/DiagnosticDescriptorBuilderTest.cs [hunk 1]
```diff
@@ -57,5 +61,75 @@ public void GetDescriptor_SetsIsEnabledByDefaultToTrue()
...
+        {
+            // Arrange
+            var diagnosticId = "foo";
+            var mockedResourceManager = CreateMockedResourceManager(diagnosticId, true);
+
+            // Act
+            var result = DiagnosticDescriptorBuilder.GetDescriptor(diagnosticId, "", IdeVisibility.Hidden, mockedResourceManager);
+
+            // Assert
+            result.CustomTags.Should().ContainInOrder(DiagnosticTagsHelper.SonarWayTag, WellKnownDiagnosticTags.Unnecessary);
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/DiagnosticDescriptorBuilderTest.cs]: This check is over-specific; we don't care about the order, only that those and only those tags are present.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs [hunk 2]
```diff
@@ -29,19 +30,32 @@ namespace SonarAnalyzer.Helpers
     public class SonarAnalysisContext
     {
         private readonly AnalysisContext context;
-        public static Func<SyntaxTree, bool> ShouldAnalysisBeDisabled { get; set; }
+        private readonly IEnumerable<DiagnosticDescriptor> supportedDiagnostics;
+
+        /// <summary>
+        /// This delegate is only set by the VSIX when the projects have the NuGet package installed to avoid repeated behaviors.
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs]: This comment isn't accurate any more, now that we've using the same mechanism to disable specific rules.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs [hunk 3]
```diff
@@ -29,19 +30,32 @@ namespace SonarAnalyzer.Helpers
     public class SonarAnalysisContext
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs]: This class is now stateful and specific to a single analyser. It's probably worth documenting this, which would make it clearer that the list of _supportedDiagnostics_ isn't an arbitrary list, it's just the rules for a specific analyser.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/AnalysisContextExtensions.cs [hunk 4]
```diff
@@ -60,7 +60,8 @@ public static void ReportDiagnosticWhenActive(this CodeBlockAnalysisContext cont
 
         private static void ReportWhenNotSuppressed(SyntaxTree tree, Diagnostic diagnostic, Action<Diagnostic> report)
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/AnalysisContextExtensions.cs]: This method name is misleading now that more than the suppression state being taken into account.

Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/AnalysisContextExtensions.cs]: WDYT of `TryReportDiagnostic`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs (hunk 3 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs]: It was already stateful (was taking the roslyn context). But I will add a comment.

**Pull request closed without merging**